src/components/OntologyBrowser/SelectionTree.tsx

```diff
@@ -78,30 +79,65 @@ export class SelectionTree extends Component<SelectionTreeProps, SelectionTreeSt
     selectedKeys.indexOf(eventKey.toString()) !== -1;
 
   onChange = (e: React.ChangeEvent<HTMLInputElement>, treeData: TreeNode[]) => {
-    treeData.forEach(node => this.searchInTree(e.target.value, node));
-    this.setState({ treeData });
+    const hits: string[] = []
+
+    if(e.target.value.length >= MIN_SEARCH_TEXT_LENGTH){
+      treeData
+        .forEach(node => this.searchInTree(e.target.value, node, hits))
+
+      this.setState({
+        treeData:treeData,
+        expandedKeys: hits,
+      })
+    } else {
+      treeData
+        .forEach(node => this.unhideAll(node))
+      this.setState({
+        treeData: treeData,
+        expandedKeys: getInitialKeysForExpand(treeData),
+      });
+    }
   };
 
-  searchInTree = (searchText: string, treeNode: TreeNode) => {
-    const regex = new RegExp(searchText, 'gi');
-    const text = treeNode.title as string;
+  searchInTree = (searchText: string, treeNode: TreeNode, hitTreeNodes: string[] = []) => {
+    let regex = new RegExp('\\b(\\w*' + searchText + '\\w*)\\b', 'gi')
+    const text = treeNode.text;
     const result = text.search(regex) >= 0;
     let match = searchText === '' || result;
 
     if (treeNode.children.length > 0) {
       let matchChild = searchText === '' || false;
       treeNode.children.forEach((child: TreeNode) => {
-        if (this.searchInTree(searchText, child)) {
+        if (this.searchInTree(searchText, child, hitTreeNodes)) {
           matchChild = true;
         }
       });
       match = matchChild || match;
     }
-
     treeNode.hidden = !match;
+
+    if(!treeNode.hidden) {
+      hitTreeNodes.push(treeNode.key)
+      if(result) {
+        const [ before, hit, after ] = treeNode.text.split(regex)
```

Comment:
Is it ok if there is no before or after?

Reply:
No, its not ok. But it should never happen. There is a check before hand `if(result) {...}`, therefore there is a match with the regex. If so, before and after cannott be `undefined`. One might be empty strings if the regex match is at the beginning/end or both empty is the full term is matched by the regex.